src/vsr/replica.zig

```diff
@@ -2923,26 +3024,70 @@ pub fn ReplicaType(
-            assert(self.view >= self.log_view);
             assert(self.view >= self.view_durable());
             assert(self.log_view >= self.log_view_durable());
-
-            assert(command != .do_view_change or self.log_view < self.view);
-            assert(command != .start_view or self.log_view == self.view);
+            assert((self.log_view < self.view) == (command == .do_view_change));
+            assert((self.log_view == self.view) == (command == .start_view));
+            assert(self.log_view < self.view or self.replica == self.primary_index(self.view));
 
             const message = self.message_bus.get_message();
             defer self.message_bus.unref(message);
 
-            const headers = self.create_view_change_headers();
-            assert(headers.array.len > 0);
-            assert(headers.array.get(0).op == self.op);
+            switch (command) {
+                // The DVC headers are already up to date, either via:
+                // - transition_to_view_change_status(), or
+                // - superblock's vsr_headers (after recovery).
+                .do_view_change => {},
+                .start_view => {
+                    self.view_headers.array.len = 0;
+
+                    var op = self.op + 1;
+                    while (op > 0 and
+                        self.view_headers.array.len < constants.view_change_headers_suffix_max)
+                    {
+                        op -= 1;
+                        self.view_headers.array.appendAssumeCapacity(self.journal.header_with_op(op).?.*);
+                    }
+                    assert(self.view_headers.array.len + 2 + 1 <= constants.view_change_headers_max);
+
+                    // Determine the consecutive extent of the log that we can help recover.
+                    // This may precede op_repair_min if we haven't had a view-change recently.
+                    const range_min = (self.op + 1) -| constants.journal_slot_count;
+                    const range = self.journal.find_latest_headers_break_between(range_min, self.op);
+                    const op_min = if (range) |r| r.op_max + 1 else range_min;
+                    assert(op_min <= op);
+                    assert(op_min <= self.op_repair_min());
+                    assert(op_min <= self.op_checkpoint_trigger_previous());
+
+                    // The SV includes headers corresponding to:
+                    // - the triggers for preceding checkpoints (as many as we have and can help
+                    //   repair, which is at most 2), and
+                    // - the oldest op that we can help repair.
+                    for ([_]u64{
+                        self.op_checkpoint_trigger() -|
+                            (constants.journal_slot_count - constants.lsm_batch_multiple),
+                        self.op_checkpoint_trigger() -|
+                            (constants.journal_slot_count - constants.lsm_batch_multiple) * 2,
+                        op_min,
```

Comment:
The last (i.e. oldest) header in the SV is special.

`on_start_view` knows that if the oldest header in the SV has op `X`, then the primary can help it repair all ops `≥X`. (Or at least it could when the SV was generated). `on_start_view` uses this to determine when it needs to fall back to state synchronization.

Reply:
:thinking: don't yet fully understand this bit, but this last op feels suspect to me. My understanding is that this'll be the first op to be overwritten by an incoming request. So, it feels a bit useless to send it out --- there's a high chance that we _won't_ actually remeber this op by the time the message reaches the destination.

Reply:
It corresponds to: https://github.com/tigerbeetledb/tigerbeetle/blob/ff1b61f29f4e8c1fc78edc51f88308ead10fd0b8/src/vsr/replica.zig#L1485-L1508 on the `on_start_view` side. That is, the backup uses it to determine when it is definitely too far behind the primary and must transition to state transfer.

> there's a high chance that we won't actually remeber this op by the time the message reaches the destination.

True, when the cluster is healthy. Though after a view-change, there's a good chance that the cluster isn't healthy.

:thinking: If we go to fetch that header and it isn't available, then we will keep trying & maybe even get stuck right now. Maybe `command=commit` heartbeats should include a "minimum repairable message" which backups can use to transition to state transfer.

Reply:
>on_start_view knows that if the oldest header in the SV has op X, then the primary can help it repair all ops ≥X

Ok, I think this is backwards a bit, what we actually get here is:

> if the oldest header in the SV has op X, then it's not possible to repare any op `<x`

So, when a replica receives an SV with oldest op X, and it doesn't have X itself, then it knows it should switch to state sync. If it has op X, it might attempt to request missing prepares.

Reply:
>heartbeats should include a "minimum repairable message" which backups can use to transition to state transfer.

I think a backup can infer that from `commit`? if heartbeat commit is at least a wrap ahead of our commit_min, we should switch to sync (give or take a pipeline worth of messages).

Reply:
> I think a backup can infer that from commit? if heartbeat commit is at least a wrap ahead of our commit_min, we should switch to sync (give or take a pipeline worth of messages).

Good point; updated. `repair` now uses `commit_max`  to identify messages that it cannot possibly repair, and trigger state sync.